### ISSUE TITLE
ALPN requirement mismatch?

### ISSUE THREAD

**Issue by thomas-fossati** [yaronf/I-D#370]
https://datatracker.ietf.org/doc/review-ietf-uta-rfc7525bis-07-secdir-lc-kaduk-2022-06-01/

```
Section 3.8

   TLS implementations (both client- and server-side) MUST support the
   Application-Layer Protocol Negotiation (ALPN) extension [RFC7301].

   In order to prevent "cross-protocol" attacks resulting from failure
   to ensure that a message intended for use in one protocol cannot be
   mistaken for a message for use in another protocol, servers SHOULD
   strictly enforce the behavior prescribed in Section 3.2 of [RFC7301]:
   "In the event that the server supports no protocols that the client
   advertises, then the server SHALL respond with a fatal
   no_application_protocol alert."  [...]

It feels a bit strange to have a "MUST support [RFC7301]" statement and
follow it up with only a "SHOULD strictly enforce [part of RFC 7301]".  If
you MUST do it, don't you need to always do it, not just sometimes (as
SHOULD would have it)?

   Protocol developers are strongly encouraged to register an ALPN
   identifier for their protocols.  This applies to new protocols, as
   well as well-established protocols.

I'd consider following this with a clause "but strict enforcement of ALPN
usage may not be feasible when the ALPN identifier is registered for a
well-established protocol".
```

cc @kaduk

**Comment by yaronf** [comment]
The second point is not clear. @kaduk Do you mean "when the identifier is *not* registered"?

**Comment by yaronf** [comment]
Closed via #387 .

**Comment by kaduk** [comment]
[belated confirmation that #387 did what I wanted -- when the ALPN identifier is registered after the protocol is already deployed, the existing deployments will not use it, so a server that attempts to require that ALPN is always used would break new clients.  You need some additional information/constraint, e.g., about only updated clients being used to talk to a given server, in order to safely let the server require the ALPN to be used.]
